native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java

```diff
@@ -528,6 +530,26 @@ private static void configureAgent(Project project,
         nativeImageOptions.getConfigurationFileDirectories().from(files);
     }
 
+    private static Provider<Directory> agentOutputDirectoryFor(Project project, NativeImageOptions nativeImageOptions, TaskProvider<? extends JavaForkOptions> instrumentedTask) {
```

Comment:
This isn't correct, as you are effectively eagerly reading the provider (`orElse(AGENT_...`), which defeats the laziness.

Reply:
It's my first time working with gradle, so I might be mistaken, but as far as I can tell, there are no lazy input values to the computation of the "outputDirUnresolved" variable (the part of the code ending with `orElse(AGENT_...`). All the values involved come from the plugin's configuration which I expect to be constant from start to finish. And since `.getBuildDirectory().dir(outputDirUnresolved)` returns a provider I would assume that whatever it does with outputDirUnresolved will be done lazily.

In case I'm missing something, I have pushed another version where the whole computation is wrapped into one of the Provider implementations I found in gradle's plugin code.

Reply:
:x: this isn't correct either. You should not implement your own provider type. I will take a stab at it once the CLA issues are fixed.